### PR TITLE
feat: add support for jwt on api

### LINKED ISSUE #168: Improve token auth
#### My reason:
Current implement make token expire after 24 hours, it's impossible to use with CI or automatic deploy.
I think we can let user choose token expire time or never expire with reset feature.

Use ([JTW](https://jwt.io) may better than origin code, any idea? @juanpicado

### PR DESCRIPTION
**Pull request by juanpicado**
**Type:** feat
**Description:**

This PR aims to deprecate AES token generator but without replace it and allow to use JWT as alternative on API.  (https://github.com/verdaccio/verdaccio/issues/168#issuecomment-394149624)

Provides customise JWT signature and verification based on 
https://github.com/auth0/node-jsonwebtoken#usage

```
security:
  api:
    jwt:
      sign:
        expiresIn: 60d
        notBefore: 1
  web:
    sign:
      expiresIn: 7d
```

`.npmrc`
```
//localhost:4873/:_authToken=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoianVhbjExIiwiZ3JvdXBzIjpbImp1YW4xMSIsIiRhbGwiLCIkYXV0aGVudGljYXRlZCIsIkBhbGwiLCJAYXV0aGVudGljYXRlZCIsImFsbCJdLCJyZWFsX2dyb3VwcyI6WyJqdWFuMTEiXSwiZ3JvdXAiOlsianVhbjExIl0sImlhdCI6MTUzNDAwOTkzOSwibmJmIjoxNTM0MDA5OTM5LCJleHAiOjE1MzQ2MTQ3Mzl9.FPQ__wb-CIT2Gnti0fMyBw6XK20Zv8xuuZBqVVaCr7s
```

Resolves #168 https://github.com/verdaccio/verdaccio/issues/729

🆕  enable JWT by default on API.

### CODE REVIEW

File: src/api/web/index.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ const spliceURL = require('../../utils/string').spliceURL;
 module.exports = function(config, auth, storage) {
   Search.configureStorage(storage);
 
-  router.use(auth.webUIJWTmiddleware());
+  // router.use(auth.webUIJWTmiddleware());
```
Comment (juanpicado) [inline src/api/web/index.js]: reminder: this is odd

File: src/lib/auth-utils.js [hunk 2]
```diff
@@ -36,3 +87,162 @@ export function getDefaultPlugins() {
...
+    const credentials = convertPayloadToBase64(token).toString();
+
+    return credentials;
+  } else if (scheme.toUpperCase() === TOKEN_BEARER.toUpperCase()) {
+    const tokenAsBuffer = convertPayloadToBase64(token);
+    const credentials = aesDecrypt(tokenAsBuffer, secret).toString('utf8');
+
+    return credentials;
+  } else {
+    return;
```
Comment (ayusharma) [inline src/lib/auth-utils.js]: No need of `return`

File: src/lib/auth-utils.js [hunk 3]
```diff
@@ -36,3 +87,162 @@ export function getDefaultPlugins() {
...
+
+export function parseAuthTokenHeader(authorizationHeader: string): AuthTokenHeader {
+  const parts = authorizationHeader.split(' ');
+  const [scheme, token] = parts;
+
+  return {scheme, token};
+}
+
+export function parseBasicPayload(credentials: string): BasicPayload {
+  const index = credentials.indexOf(':');
```
Comment (ayusharma) [inline src/lib/auth-utils.js]: `.includes`

File: src/lib/auth-utils.js (hunk 2 above)
Comment (ayusharma) [reply src/lib/auth-utils.js]: may be you want it more readable..

File: src/lib/auth-utils.js [hunk 4]
```diff
@@ -36,3 +87,162 @@ export function getDefaultPlugins() {
...
+    }
+
+    return parsedCredentials;
+  } else {
+    const {scheme, token} = parseAuthTokenHeader(authorizationHeader);
+
+    if (_.isString(token) && scheme.toUpperCase() === TOKEN_BEARER.toUpperCase()) {
+        return verifyJWTPayload(token, secret);
+    } else {
+      return;
```
Comment (ayusharma) [inline src/lib/auth-utils.js]: same here.. return

File: src/lib/auth-utils.js [hunk 5]
```diff
@@ -36,3 +87,162 @@ export function getDefaultPlugins() {
...
+
+  if (_.isNil(config.security) === false) {
+    return _.merge(defaultSecurity, config.security);
+  }
+
+  return defaultSecurity;
+}
+
+export function getAuthenticatedMessage(user: string): string {
+  return `you are authenticated as '${user}'`;
```
[the remark this answers is not in the document]
Comment (juanpicado) [reply src/lib/auth-utils.js]: Yes, it's intended. Don't ask me why :-) ... it was there already and unless I know what it does i cannot remove it.

File: src/lib/auth-utils.js (hunk 4 above)
Comment (juanpicado) [reply src/lib/auth-utils.js]: it was mostly for debugging. Removed

**Comment by juanpicado** [comment]
Thanks @priscilawebdev @ayusharma for CR, this is a nice feature.

**Pull request merged**